### PR TITLE
CON-212 Add timestamps fields to database records

### PR DESCRIPTION
**Pull request by koitoror**
#### Description
This Pull request add created_date and updated_date fields to the database all relevant database records

#### How should this be manually tested?
1. Clone the repo: `git clone https://github.com/andela/mrm_api.git`
2. Setup project according to Readme.md
3. checkout to branch `story/CON-212-add-created-and-updated-timestamps-to-database-records`

**How should this be manually tested?**

- use the query below to get all questions 
```
query {
  allQuestions {
   startDate
   dateUpdated
 }
}
```
- use the mutation below to update locations

```
mutation {
        changeUserLocation(email: "daniel.kamar@andela.com", locationId: 2) {
          user {
            name
            location
          }
        }
      }
```

#### Type of Change
- [x] New feature (non-breaking change which adds functionality

#### How has this been tested
- [x] Unit tests
- [x] Integration tests

#### Checklist
- [x] My changes generate no new warnings
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [x] Existing unit tests pass locally with my changes
- [x] Implementation works according to expectations

#### JIRA
[CON-212](https://andela-apprenticeship.atlassian.net/browse/CON-212)
Screenshot 2019-08-12 at 07 16 40

### CODE REVIEW

File: api/devices/models.py [hunk 1]
```diff
@@ -13,6 +13,7 @@ class Devices(Base, Utility):
     name = Column(String, nullable=False)
     device_type = Column(String, nullable=False)
     date_added = Column(DateTime, nullable=False)
+    date_updated = Column(DateTime, nullable=False)
```
Comment (vic3king) [inline api/devices/models.py]: I think you need to set a default value for this field. e.g `date_updated = Column(String, default=datetime.datetime.now)`

Comment (koitoror) [reply api/devices/models.py]: @vic3king : that presents issues on testing, therefore, I have opted set nullable to true (nullable=True) and my justification for this is that whoever gets tasked to update a specific record  could manipulate nullable to false (nullable=False), set a default value if any, come up with a separate update migration file to handle that scope and also fix the tests.

File: api/response/models.py [hunk 2]
```diff
@@ -27,6 +27,7 @@ class Response(Base, Utility):
     question_type = Column(Enum(QuestionType), default="rate")
     response = Column(postgresql.ARRAY(sa.String()))
     created_date = Column(DateTime, nullable=False)
+    updated_date = Column(DateTime, nullable=True)
```
Comment (rajeman) [inline api/response/models.py]: @koitoror Is there any reason why this timestamp column has a different name?

Comment (koitoror) [reply api/response/models.py]: @rajeman : unless for uniformity only, let me fix!

File: helpers/database.py [hunk 3]
```diff
@@ -14,5 +14,12 @@
                                          autoflush=False,
                                          bind=engine))
 
-Base = declarative_base()
-Base.query = db_session.query_property()
+DeclarativeBase = declarative_base()
+DeclarativeBase.query = db_session.query_property()
+
+
+class Base(DeclarativeBase):
```
Comment (IssaIan) [inline helpers/database.py]: You can add docstrings here @koitoror to explain what the class does.

**Comment by vic3king** [comment]
> Since these columns will be on all models and all the models make use of the `Base` object, I think you can make the `Base` object an `abstract model class` that has the two columns. Also, they should have default timestamp values and should not be nullable.

@koitoror Has this been worked on?

**Comment by koitoror** [comment]
> > Since these columns will be on all models and all the models make use of the `Base` object, I think you can make the `Base` object an `abstract model class` that has the two columns. Also, they should have default timestamp values and should not be nullable.
> 
> @koitoror Has this been worked on?

Yes

File: alembic/versions/27a555f0c123_add_create_and_update_timestamps_on_all_.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+"""add create and update timestamps on all records
+
+Revision ID: 27a555f0c123
+Revises: 20dd40a4c970
```
Comment (rajeman) [inline alembic/versions/27a555f0c123_add_create_and_update_timestamps_on_all_.py]: @koitoror You will need to update the revision of this alembic file as it is no longer up to date with what we have on develop.

**Comment by rajeman** [comment]
@koitoror You have not made the change. Your alembic version is out of date.

File: alembic/versions/27a555f0c123_add_create_and_update_timestamps_on_all_.py (hunk 4 above)
Comment (koitoror) [reply alembic/versions/27a555f0c123_add_create_and_update_timestamps_on_all_.py]: Done

**Comment by koitoror** [comment]
> Nice Implementation @koitoror. What is going to be done on the columns `date_added` on `devices model and `created_date`on`responses model`now that they are the same with`date_created` column

the `date_created` and `date_updated` timestamp fields that are being introduced on all database tables are strictly for watching/noting/checking the times/dates when an entry is added or and watching for when an entry is updated and have no other purpose aside of that. therefore they would co-exist, look/feel similar but their purposes are different.

**Pull request merged**